core/state-machine/src/changes_trie/build.rs

```diff
@@ -38,45 +39,114 @@ pub fn prepare_input<'a, B, H, Number>(
+		let dig_iter = children_digest_input.remove(&child_index);
+		children_digest.push((
+			child_index,
+			Some(ext_iter).into_iter().flatten()
+				.chain(dig_iter.into_iter().flatten()),
+		));
+	}
+	for (child_index, dig_iter) in children_digest_input.into_iter() {
+		children_digest.push((
+			child_index,
+			None.into_iter().flatten()
+				.chain(Some(dig_iter).into_iter().flatten()),
+		));
+	}
+
+	Ok((
+		extrinsics_input.chain(digest_input),
+		children_digest,
+	))
+}
 /// Prepare ExtrinsicIndex input pairs.
 fn prepare_extrinsics_input<'a, B, H, Number>(
 	backend: &'a B,
 	block: &Number,
 	changes: &'a OverlayedChanges,
+) -> Result<(
+		impl Iterator<Item=InputPair<Number>> + 'a,
+		BTreeMap<ChildIndex<Number>, impl Iterator<Item=InputPair<Number>> + 'a>,
+	), String>
+	where
+		B: Backend<H>,
+		H: Hasher + 'a,
+		Number: BlockNumber,
+{
+
+	let mut children_keys = BTreeSet::<Vec<u8>>::new();
+	let mut children_result = BTreeMap::new();
+	for (storage_key, _) in changes.prospective.children.iter()
+		.chain(changes.committed.children.iter()) {
+		children_keys.insert(storage_key.clone());
+	}
+	for storage_key in children_keys {
+		let child_index = ChildIndex::<Number> {
+			block: block.clone(),
+			storage_key: storage_key.clone(),
+		};
+	
+		let iter = prepare_extrinsics_input_inner(backend, block, changes, Some(storage_key))?;
+		children_result.insert(child_index, iter);
+	}
+
+	let top = prepare_extrinsics_input_inner(backend, block, changes, None)?;
+
+	Ok((top, children_result))
+}
+	
+fn prepare_extrinsics_input_inner<'a, B, H, Number>(
+	backend: &'a B,
+	block: &Number,
+	changes: &'a OverlayedChanges,
+	storage_key: Option<Vec<u8>>,
 ) -> Result<impl Iterator<Item=InputPair<Number>> + 'a, String>
 	where
 		B: Backend<H>,
 		H: Hasher,
 		Number: BlockNumber,
 {
-	changes.committed.top.iter()
-		.chain(changes.prospective.top.iter())
+	let (committed, prospective) = if let Some(sk) = storage_key.as_ref() {
+		(changes.committed.children.get(sk), changes.prospective.children.get(sk))
+	} else {
+		(Some(&changes.committed.top), Some(&changes.prospective.top))
+	};
+	committed.iter().flat_map(|c| c.iter())
+		.chain(prospective.iter().flat_map(|c| c.iter()))
 		.filter(|( _, v)| v.extrinsics.is_some())
 		.try_fold(BTreeMap::new(), |mut map: BTreeMap<&[u8], (ExtrinsicIndex<Number>, Vec<u32>)>, (k, v)| {
 			match map.entry(k) {
 				Entry::Vacant(entry) => {
 					// ignore temporary values (values that have null value at the end of operation
 					// AND are not in storage at the beginning of operation
-					if !changes.storage(k).map(|v| v.is_some()).unwrap_or_default() {
+					let existing = if let Some(sk) = storage_key.as_ref() {
+						changes.child_storage(sk, k)
+					} else {
+						changes.storage(k)
+					};
+					if !existing.map(|v| v.is_some()).unwrap_or_default() {
 						if !backend.exists_storage(k).map_err(|e| format!("{}", e))? {
```

Comment:
If `storage_key.is_some()`, there should be a call to `backend.exists_child_storage()`

Reply:
pretty bad one. I remember when looking at this part of code that this query should be expensive, I guess there may be something doable to have the info in the transaction without querying the backend (would probably require additional query of backend during execution in some case).

Reply:
We have had this discussion in #2865 - it was an (unsuccessful) attempt to extend this check to non-temporary values. IMO that's what state cache actually handles - if value has been read from trie during execution, then it'll be read from in-memory cache for the second time. And if it wasn't, then there's no other way, than to perform this read.

Reply:
I also guess that running this backend check during block execution, in case of invalid block, is also an unnecessary operation.
Maybe if overlay was aware of accessed data, it could conditionally store this in the transaction, but that would be similar as putting some access cache in the overlayed_change (which may be interesting to avoid some map access (we already need to check change for get_storage) but would be some substantial design change).